app/core/service/ProxyModeService.ts

```diff
@@ -0,0 +1,365 @@
+import { InternalServerError, ForbiddenError, HttpError } from 'egg-errors';
+import { SingletonProto, AccessLevel, Inject } from '@eggjs/tegg';
+import {
+  EggContextHttpClient,
+} from 'egg';
+import { calculateIntegrity } from '../../common/PackageUtil';
+import { getScopeAndName } from '../../common/PackageUtil';
+import { downloadToTempfile } from '../../common/FileUtil';
+import { NPMRegistry, RegistryResponse } from '../../common/adapter/NPMRegistry';
+import { UserRepository } from '../../repository/UserRepository';
+import { AbstractService } from '../../common/AbstractService';
+import { RegistryManagerService } from './RegistryManagerService';
+import { PackageManagerService } from './PackageManagerService';
+import { PackageRepository } from '../../repository/PackageRepository';
+import { UserService } from './UserService';
+import { CacheService } from './CacheService';
+import { User } from '../entity/User';
+import { rm, readFile } from 'fs/promises';
+import { NFSAdapter } from '../../common/adapter/NFSAdapter';
+import { Registry } from '../entity/Registry';
+import { RegistryType } from '../../common/enum/Registry';
+import { PROXY_MODE_CACHED_PACKAGE_DIR_NAME } from '../../common/constants';
+import { DIST_NAMES } from '../entity/Package';
+
+@SingletonProto({
+  accessLevel: AccessLevel.PUBLIC,
+})
+export class ProxyModeService extends AbstractService {
+  @Inject()
+  private readonly httpclient: EggContextHttpClient;
+  @Inject()
+  private readonly npmRegistry: NPMRegistry;
+  @Inject()
+  private readonly userRepository: UserRepository;
+  @Inject()
+  private readonly packageRepository: PackageRepository;
+  @Inject()
+  private readonly registryManagerService: RegistryManagerService;
+  @Inject()
+  private readonly packageManagerService: PackageManagerService;
+  @Inject()
+  private readonly userService: UserService;
+  @Inject()
+  private readonly cacheService: CacheService;
+  @Inject()
+  private readonly nfsAdapter: NFSAdapter;
+
+  async getPackageVersionTarAndPublish(fullname: string, version: string, url: string): Promise<Buffer| null> {
+    if (this.config.cnpmcore.syncPackageBlockList.includes(fullname)) {
+      throw new ForbiddenError(`stop proxy by block list: ${JSON.stringify(this.config.cnpmcore.syncPackageBlockList)}`);
+    }
+    const requestTgzURL = `${this.npmRegistry.registry}/${url}`;
+    let tmpfile:string;
+    try {
+      ({ tmpfile } = await downloadToTempfile(this.httpclient, this.config.dataDir, requestTgzURL));
+    } catch (error) {
+      throw new InternalServerError('get uplink tgz file failed.');
+    }
+
+    const tgzBuffer = await readFile(tmpfile);
+    // FIXME: should it run in background?
+    await this.publishDownloadPackageVersionTar(fullname, version, tmpfile);
```

Comment:
this.publishDownloadPackageVersionTar(fullname, version, tmpfile).catch(err => {})

这样就代表后台执行了，记录 error 日志即可。

Reply:
直接去掉await会报
```shell
ERROR 27228 framework.TeggError: Can not read property `NFSAdapter#uploadBytes#AsyncTimer#0` because egg ctx has been destroyed [ https://eggjs.org/zh-cn/faq/TEGG_read_after_ctx_destroyed ]
```
tegg的文档上写用`SingletonProto`的对象不会释放但实际的表现和之前用`ContextProto`是一样的，是否还是需要安装`@eggjs/tegg-background-task`？或者通过任务的方式发布？